File: 백준/Gold/1863. 스카이라인 쉬운거/스카이라인 쉬운거.py
import sys

n = int(sys.stdin.readline())

skyline = []
for i in range(n):
    skyline.append(list(map(int, sys.stdin.readline().split())))

has_height = [0] # stack

answer = 0
for idx, h in skyline:
    if has_height[-1] < h:
        answer += 1
        has_height.append(h)
    else:
        while has_height[-1] > h:
            has_height.pop()
        
        if has_height[-1] < h:
            answer += 1
            has_height.append(h)

print(answer)

File: 백준/Gold/1863. 스카이라인 쉬운거/스카이라인 쉬운거.cc
#include <string>
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

int main(){
    int n;
    cin >> n;
    
    vector<vector<int>> skyline(n, vector<int>(2, 0)); // index, height
    for(int i = 0; i < n; i++){
        cin >> skyline[i][0];
        cin >> skyline[i][1];
    }
    
    stack<int> stk;
    stk.push(0);
    
    int answer = 0;
    for(int i = 0; i < n; i++){
        int h = skyline[i][1];
        if(stk.top() < h){
            answer++;
            stk.push(h);
        }
        else{
            while(stk.top() > h) stk.pop();
            
            if(stk.top() < h){
                answer++;
                stk.push(h);
            }
        }
    }
    cout << answer << endl;
    
    return 0;
}